### ISSUE TITLE
Backpressure on fetch integrated with Streams

### ISSUE THREAD

**Issue by yutakahirano** [slightlyoff/ServiceWorker#452]
Imagine ReadaleStream is integrated with fetch. The most naive way is to add stream() method to Body interface.

    interface Body {
        ...
      Promise<ReadableStream> stream();
    };

Streams is an API that enables the data producer to know that the consumer doesn't want the data and stop producing data. On the other hand, other data types (text, blob, ...) don't have such feature.
My question is, should the loader start loading body while the Body representation is undetermined?

    fetch(request).then(function(res) {
      return wait(longtime).then(function() {
        return res.stream();
      });
    }).then(function(stream) {
      // Here we have a readable stream.
    });

The above code waits *longtime* and then demonstrates that it wants to get the body data via ReadableStream interface. Should the loader load the body data while waiting and use the internal buffer unlimitedly, or stop loading while the representation is undetermined?

We can provide that information explicitly when calling *fetch*, but I don't know if it's a good API.

**Comment by domenic** [comment]
OK I just got back into all this and had to re-read the thread because I forgot what we eventually decided on for `bodyUsed`.

Given @yutakahirano's above comment, can we rename it to `bodyLocked`? Or maybe we should just expose it on the stream directly as `body.isLocked`?

**Comment by yutakahirano** [comment]
@domenic  and I talked a bit at https://github.com/yutakahirano/fetch-with-streams/issues/13. Now we should revisit `.bodyUsed` and `.locked`, because they are different concepts.

- The current `.bodyUsed` specified at https://fetch.spec.whatwg.org/ represents whether some functions such as `text()` and `json()` are called. For example, `res.text().then(() => res.text())` will be rejected, because `.bodyUsed` is turned into true at the first `text()` call and it will never become false. Note that `.bodyUsed` becomes true even when no data is read - Think about the case when we call `text()` on a response that has an empty body.
- A stream can be locked. The lock will be released when explicitly indicated, the stream gets closed or the stream gets errored. `text()` acquires a lock, but it releases the lock when the stream is closed or errored.
- There is another concept: whether body data is read. Let me call it `.bodyConsumed` for now. It becomes true when some data is read via ReadableStream.prototype.read. `res.text()` will turn this flag into true when the data is actually read (not when the function is called). Note that `res.text()` on a response with an empty body doesn't make this flag true.

What concept do we need? Surely `.locked` is needed, but it seems insufficient. `.bodyConsumed` is good from some security POV (because it is useful to guarantee that the response is not partially consumed). The problem is `.bodyConsumed` is hard to implement. @domenic's [idea](https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-56049602) needs a custom ReadableByteStream for FetchAPI, and a custom ExclusiveStreamReader for Fetch API because the standard ExclusiveStreamReader doesn't care about the customized `read()`.
The existing `.bodyUsed` concept doesn't work with `.body` attribute, because we can freely read from stream even when `.bodyUsed` is true.

How about @tyoshino's [old idea](https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-56145980)? It is simple, and we can guarantee that the body is not partially consumed when the flag is true. It is easy to implement and easy to understand. Currently I don't have a better idea.

**Comment by domenic** [comment]
>  @domenic's idea needs a custom ReadableByteStream for FetchAPI, and a custom ExclusiveStreamReader for Fetch API because the standard ExclusiveStreamReader doesn't care about the customized read().

I don't understand why? It needs a custom underlying source, but every type of stream has a custom underlying source, so a fetch body stream should too.

One way to phrase it (that's a bit more explicit than my code that you link to) is that the underlying source sets `bodyUsed` to `true` anytime it calls `enqueue()`.

> How about @tyoshino's old idea?

I am not sure whether that idea gives any value to developers. It just tells them that somebody once accessed the `body` property---not whether any data has been read from it.

I think we need to step back and ask what the purpose of `bodyUsed` is? How do we see developers using it?

**Comment by yutakahirano** [comment]
>> @domenic's idea needs a custom ReadableByteStream for FetchAPI, and a custom ExclusiveStreamReader for Fetch API because the standard ExclusiveStreamReader doesn't care about the customized read().

> I don't understand why? It needs a custom underlying source, but every type of stream has a custom underlying source, so a fetch body stream should too.
> 
> One way to phrase it (that's a bit more explicit than my code that you link to) is that the underlying source sets bodyUsed to true anytime it calls enqueue().

Let me confirm:

```
var response;
var p = fetch(...).then((res) => {
  response = res;
  return res.body.ready;
}).then(() => 
  // This is not generally true, but please assume this holds in this example.
  assert(response.body.state === 'readable');

  return response.text();
});
```

Will `p` be resolved, or rejected? `text()` is called after `enqueue()` is called, so it should be rejected, right?

**Comment by domenic** [comment]
Oh I see, you are right. Enqueue is not the right place to put that. Interesting that it seems like it would work if we marked bodyUsed inside a `ReadableByteStream`'s underlying source's `readInto`, but it will not work with a plain `ReadableStream`. Do you agree with that?

If that is right, maybe we should not have a `bodyUsed` until we implement the `ReadableByteStream` interface?

**Comment by annevk** [comment]
To be clear, I don't really care what we do here, but I'd like the invariants in the specification to be preserved (around the "used flag").

We expose `bodyUsed` so developers can detect whether serialize methods (`text()` et al), `clone()` methods, and constructing a `Request` based on another `Request`, might fail. Not exposing it leads to try/catch or worse (serialize methods return promises).

**Comment by yutakahirano** [comment]
Sorry, I misunderstood ReadableByteStream.

As you say, we can observe the underlying source's `readInto`, so we definitely can know when `read` or `readInto` is called.

 - Body mixin's `used flag` is set when stream's `readInto` is called.
   - But, `used flag` can be incorrect when the stream is locked to an exclusive lock, to give an optimization chance to implementations.
 - Remove `Body.bodyUsed`. Because `used flag` is valid only when stream is not locked.

Can we define like this? This introduces incompatible changes.

 - `Body.bodyUsed` is removed.
 - Reading an empty body does not set `used flag`. For example, if `res` is a response with an empty body, `res.text().then(() => res.text())` is resolved with an empty string.

**Comment by yutakahirano** [comment]
> To be clear, I don't really care what we do here, but I'd like the invariants in the specification to be preserved (around the "used flag").

@annevk, I would like know if the followings are important:

- `Body.bodyUsed` is exposed to JavaScript.
- Once `used flag` becomes true, it will never be false.
- Calling `text()` twice will fail, even when the body is empty.

**Comment by annevk** [comment]
Why would the "used flag" be unset? I could imagine giving in on the other two, depending on the final picture.

**Comment by yutakahirano** [comment]
If we set `used flag` based on whether the body data is consumed, the exact value should not be observable when body is locked to a reader. That is why I remove `Body.bodyUsed` in 
https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70622279. If we must expose the value, we may set `used flag` when body is locked. If no read call is made and stream is unlocked, `used flag` will become false again.

**Comment by annevk** [comment]
Note that one thing you would have to redesign is that if you construct `Request` from a `Request` that has a body, we move the body to the new object and set the "used flag" on the old object.

**Comment by wanderview** [comment]
@yutakahirano 

Just to clarify one thing:


> -    The current .bodyUsed specified at https://fetch.spec.whatwg.org/ represents whether some functions such as text() and json() are called. For example, res.text().then(() => res.text()) will be rejected, because .bodyUsed is turned into true at the first text() call and it will never become false. Note that .bodyUsed becomes true even when no data is read - Think about the case when we call text() on a response that has an empty body.
> -    A stream can be locked. The lock will be released when explicitly indicated, the stream gets closed or the stream gets errored. text() acquires a lock, but it releases the lock when the stream is closed or errored.
> -    There is another concept: whether body data is read. Let me call it .bodyConsumed for now. It becomes true when some data is read via ReadableStream.prototype.read. res.text() will turn this flag into true when the data is actually read (not when the function is called). Note that res.text() on a response with an empty body doesn't make this flag true.

The .bodyUsed flag is also set when the UA drains the body.  For example, Cache.put() will drain the bodies for both the Request and Response objects.  Cache.put(req, res) will result in both req.bodyUsed and res.bodyUsed being set to true.

Setting these used flags occurs at the beginning of the put(), before any data is read yet.  I believe we did this so that the UA can read the data on another thread without worrying about racing with content for the body data.  See steps 3.2 and 5.2.

  https://slightlyoff.github.io/ServiceWorker/spec/service_worker/index.html#cache-put

Just wanted to note that as it didn't seem directly covered by your other cases.

**Comment by yutakahirano** [comment]
@wanderview ReadableStream has "exclusive reader" concept https://streams.spec.whatwg.org/#reader-class that corresponds to `.locked` in my text. If we use `.bodyConsumed` as `used flag`, we also use `.locked`. Then any other reader cannot read the body, so I think the existing behavior will be preserved (though the spec needs to be updated).

Note: I'm not fully convinced to use `.bodyConsumed`. https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-56145980 and https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-56147654 look also reasonable as I said before.

**Comment by domenic** [comment]
Thanks @annevk for helping remind me what `bodyUsed` was meant for. Namely, something developers can use to avoid try/catch with the convenience methods `text()` etc.

With that in mind I think `bodyUsed`, perhaps with a better name, should be true when any of the following is true:

1. The stream is currently locked
2. The stream is "closed" or "errored" but has Content-Length > 0.

I think that might be enough? That is, I think these are all of the cases where cloning or the convenience methods will fail.

A couple things to note:

- If the stream has been partially read (using `.read()`) but not read to the end, `bodyUsed` should be false, since you could still call `.text()` or clone it and get something useful.
- If `text()` is ever called on a non-empty body, then bodyUsed will be true, since either the stream will be currently locked, or it will have been closed/errored before being unlocked.
- If `text()` is called on an empty stream, then there is no effect. It is safe to call `.text()` on an empty stream as many times as you want.

Does this sound good? Did I miss anything?

**Comment by yutakahirano** [comment]
I heard that `cache.put` and `respondWith` want to assume that no bytes are read from the body, from @horo. Is my understanding correct? Is the idea shared among serviceworker guys?

**Comment by wanderview** [comment]
cache.put() checks to see if bodyUsed is set on both Request and Response.  If either is true, then it rejects.

cache.put() will also set bodyUsed on both Request and Response if there is a body to read.

So yes, currently, cache expects no bytes to have been read before the put().  I can't think of any implementation reason why we couldn't allow a partial body to be written, but it might be a big footgun for developers.

**Comment by tyoshino** [comment]
Reply to @domenic (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70718250)

> 2. The stream is "closed" or "errored" but has Content-Length > 0.

Did you want to mean this?

`stream.state === "errored" || (stream.state === "closed" && at least one byte was enqueued to the stream)`

**Comment by tyoshino** [comment]
Exhaustive list of actions and events to consider.

Actions we can take that interacts with the stream of a Response are:

- `stream.read()`
- `res.text()`
- `respondWith(res)`
- `res.clone()`
- `new Response(res)`
- `cache.put(res)`

Events that may happen on a Response's stream are:

- `stream.read()`
    - called and returned empty ArrayBuffer
    - called and returned non-empty ArrayBuffer
- `res.text()`
    - called and is reading the stream
    - called and has finished reading the stream contents
        - read nothing
        - read at least 1 byte (i.e. the stream was not empty)
- `respondWith(res)`
    - ... same as above ...
- `res.clone()`
    - ... same as above ...
- `new Response(res)`
    - ... same as above ...
- `cache.put(res)`
    - ... same as above ...

**Comment by tyoshino** [comment]
Created a matrix. Still updating.

Condition | s.read() | res.text() | respondWith(res) | new Response(res) | cache.put(res) | res.clone()
----------|---------------|------------|------------------|-------------------|---------------|------------
After stream.read() | Y | Y | Y | ? | ? | Y
res.text() ongoing | N | N | N | N | N | N
After res.text() finish | N/A | ? | ? | ? | ? | ?
After respondWith(res) | N | N | N | N | N | N
After new Response(res) | N | N | N | N | N | N
After cache.put(res) | N | N | N | N | N | N
After res.clone() | N | N | N | N | N | N

**Comment by annevk** [comment]
You mean `new Request(req)` (and you forgot `req.clone()`, though it should be roughly identical to `res.clone()`).

**Comment by tyoshino** [comment]
@annevk Oh, right. Added a note at the footnote.

**Comment by tyoshino** [comment]
It seems Anne is basically more concerned about "arbitration".

As Yutaka said in https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70782450, we were wondering if there's need for "modified flag". Maybe there's no concrete reason to ban drained request/response (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70783239) found so far?

Domenic's this idea (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70718250) is also taking care of dirtyness.

I'm wondering if we should separate these concepts (arbitration flag and modified flag) into different variables.

**Comment by yutakahirano** [comment]
> I'm wondering if we should separate these concepts (arbitration flag and modified flag) into different variables.

If we choose that way (and if `text()` uses arbitration flag only), I don't think we need @domenic's condition (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70790651). It's ok that `text()` returns a promise resolved with an empty string when the stream is closed.

**Comment by horo-t** [comment]
yutakahirano@
> I heard that cache.put and respondWith want to assume that no bytes are read from the body, from @horo. Is my understanding correct? Is the idea shared among serviceworker guys?
Sorry I am @horo-t in github :)

From the security perspective, we want to distinguish the partially consumed response.
I think:
 - The partially consumed response should be treated as the response which is created in the ServiceWorker, not the response from the server.
 - The URL of partially consumed response object should not be empty.

So my idea is:
 - response.body should be a method. (#606)
 - response.body() set bodyUsed flag, and returns Stream (ReadableStream?)
 - After response.body() was called, both cache.put(response) and FetchEvent.respondWith(rsponse) should fail.
 - If you want to read the header bytes of the response and returns to the page, you have to create a new response.

```javascript
self.addEventListener('fetch', function(event) {
  event.respondWith(
    fetch(event.request)
      .then(function(response){
          var stream = response.body();
          stream.readSomeBytes(); // This should be promise?
          return new Response(stream); // Tthis constructor is not speced yet.
        })
  });
```

**Comment by yutakahirano** [comment]
@horo-t I'm sorry for the misspelling and thank you for the detailed explanation!

**Comment by domenic** [comment]
@horo-t can you explain

> From the security perspective, we want to distinguish the partially consumed response.

This seems unfortunate. What is the attack here?

I am also confused by

> The partially consumed response should be treated as the response which is created in the ServiceWorker, not the response from the server.

What is the (user-facing) difference?

**Comment by horo-t** [comment]
> > From the security perspective, we want to distinguish the partially consumed response.

> This seems unfortunate. What is the attack here?

Sorry this security concerns was form my misunderstanding. It is solved in https://crbug.com/392409.
Please ignore my comments.

**Comment by tyoshino** [comment]
Based on Horo's reply (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70965040), I've updated the matrix (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70795093) by changing N to Y where it was marked N due to https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70782450.

**Comment by tyoshino** [comment]
It seems not only `cache.put()` but `new Response(res)` and `respondWith(res)` can also accept partially consumed data. I marked them N based on Domenic's post https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70718250. But now we can flip them?

**Comment by tyoshino** [comment]
Sorry. I didn't understand Anne's this comment (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70799372) correctly. We don't have any Response constructor that takes a Response. Horo taught me.

**Comment by annevk** [comment]
But we do have a `Request` constructor that takes a `Request`. That is what I pointed out.

**Comment by tyoshino** [comment]
Yes. Added another table for Request.

**Comment by yutakahirano** [comment]
@domenic (and probably @annevk), is it critical to prevent `text()` on a dirty closed Body (I mean, https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70718250)? Allowing it seems not a big problem to me and greatly simplifies the situation.

**Comment by domenic** [comment]
@yutakahirano what is a dirty closed body, sorry? Is that

> The stream is "closed" or "errored" but has Content-Length > 0.

?

If so I think it makes `text()` more friendly, if it rejects in those cases. What would you propose `text()` do in those cases? Return an empty string?

**Comment by yutakahirano** [comment]
@domenic yes, returning an empty string seems more consistent to me.

If we can return an empty string in that case, I would propose the following for Request.

- All actions (in @tyoshino's table) will fail when the stream is locked to a reader.
- All actions except `read()` will take an exclusive lock.
- `new Request` and `fetch()` will set `unusable flag` on the Request.
- All actions will fail when Body's `unusable flag` is set.
- `new Request` and `fetch()` will fail when Body is modified (i.e. non empty data is read from it).

These rules are moderately simple and cover @tyoshino's table requirements except the "text() on a dirty closed body" case.

(Note: `unusable flag` introduced here need careful consideration. What happens when accessing the stream of a body whose `unusable flag` is set?)

**Comment by tyoshino** [comment]
@domenic Please also take a look at my post https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70998414. It seems we no longer have to prohibit `cache.put(_, res)` and `e.respondWith(res)` on partially consumed `res`. Do you agree we can flip them to `Y` in the table?

**Comment by yutakahirano** [comment]
Discussed `unusable flag` with @tyoshino.
We thought of two solutions: Both of them don't need an explicit flag.

1. Do not release the lock even when the stream is fully consumed or an read error occurs. From the outside, `body.closed` will never be resolved or rejected. This needs an additional lock acquiring mechanism in ReadableStream, because currently specified lock is auto-released when closed or errored.
1. Tee the readable stream, pass the created one to the method (such as new Request) and error the original one. Tee is [being specified](https://github.com/whatwg/streams/issues/271) now.

**Comment by domenic** [comment]
Regarding this unusable flag stuff, I am afraid we might be losing sight of what is best for developers in favor of what is easier to implement.

@tyoshino +1 for flipping them, sounds good to me.

**Comment by tyoshino** [comment]
#606 is resolved. See https://github.com/slightlyoff/ServiceWorker/issues/606#issuecomment-71616459.

Done https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71608134. Flipped to Y.

**Comment by tyoshino** [comment]
I'd like to fill the cells with `?`. Could you guys please post your opinion about whether `clone()`-ing partially consumed (either by `text()` or `read()`) should be allowed or not.

**Comment by yutakahirano** [comment]
Cloning should be allowed as soon as possible. So when an operation is allowed in a situation, Cloning should be allowed as well. Thus my opinion is that all ? should be Y.
The result body should be marked as modified when the original one is modified.

**Comment by domenic** [comment]
In general I am in favor of leniency on allowing partially consumed bodies to be used wherever normal bodies are. So Y for me.

**Comment by tyoshino** [comment]
Thanks. Updated the table based on
- https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71641189
- https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71667825

**Comment by tyoshino** [comment]
Domenic, Yutaka,

So, finally, I'd like to confirm conclusion about `text()`, `respondWith()`, `fetch()`, etc. call on a `"closed"` Request/Response which was not empty initially.

----

IMO, it's strange if the result of `text()` call on a `"closed"` Response differs depending on the initial state of the Response. We should do either of the following:

1. Regardless of the initial state (empty or non-empty), we allow `text()` on a Response with a `"closed"` stream.
1. Have a flag in the Response indicating that the Response is "consumed", and disallow `text()` on a Response once the Response is "consumed".
1. Define "consumed" concept in the Streams API, and using it, disallow `text()` on a Response with a "consumed" stream. Rather than remembering and using the knowledge of the initial state, we define a virtual termination character `$`, put it in the queue and have some operation to consume it to explicitly mark the stream "consumed". I.e. an originally empty stream can also be "consumed".

(3) looks overkilling to me. I don't have strong opinion between (1) and (2).

**Comment by yutakahirano** [comment]
+1 for (1). Reading from a partially consumed body will be useful, and (3) is complex for users.

**Comment by domenic** [comment]
Why can't we just make `text()` always reject on `"closed"` response? E.g.

```js
function text() {
  if (this.body.state === "closed") {
    return Promise.reject(new TypeError("stream is closed!"));
  }
  // otherwise, read the stream to the end...
}
```

**Comment by yutakahirano** [comment]
In terms of an empty body, I think it should be represented as an empty String / ArrayBuffer (XHR does [so](https://xhr.spec.whatwg.org/#arraybuffer-response)). In terms of a fully consumed body, I don't see a strong reason to treat it differently from an empty body.

I don't have a specific example, but developers would want to distinguish an empty response from a communication error.

By the way, `readInto()` in ReadableByteStream may return 0. That means `read()` may return an empty ArrayBuffer and it should be ignored.

**Comment by domenic** [comment]
> In terms of an empty body, I think it should be represented as an empty String / ArrayBuffer

Agreed!

> In terms of a fully consumed body, I don't see a strong reason to treat it differently from an empty body.

Well, it kind of indicates that somebody already consumed the body, so I thought it would be good to error in that case.

But I realized that an empty string (with no chunks enqueued) is automatically closed. So there is no way to distinguish between "empty" and "read to the end", except I guess out-of-band information like Content-Length. So I see your point.

> I don't have a specific example, but developers would want to distinguish an empty response from a communication error.

Agreed with this too. Yet again JavaScript's paucity of error types strikes us :(.

---

In combination, I end up agreeing with you. So, @tyoshino's 1 from https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71785894 sounds good to me.

**Comment by tyoshino** [comment]
Thanks, Yutaka, Domenic. The table (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70795093) is already updated to reflect the agreement.

----

Now, all the columns (except for that `read()` is not allowed for a closed stream) are the same.

After making a little fix on Yutaka's algorithm in https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71579896, either of the following algorithms should work:

## Lock + `body neutered flag`

- a `Request` has a boolean `body neutered flag`
- `req.bodyUsed` = (`req.body` is locked) || `req`'s `body neutered flag` is set
- The following operations fail when `req.bodyUsed` is set. Otherwise, they acquire the lock of `req.body` and release it when done.
    - `req.arrayBuffer()`
    - `req.blob()`
    - `req.formData()`
    - `req.json()`
    - `req.text()`
- `req.clone()` fail when `req.bodyUsed` is set.
- The following operations fail when `req.bodyUsed` is set. Otherwise, set `req`'s `body neutered flag`, confiscate the underlying source and queue from `req.body` and error it.
    - `fetch(req)`
    - `new Request(req)`
    - `cache.put(req, res)`

- a `Response` has a boolean `body neutered flag`
- `res.bodyUsed` = (`res.body` is locked) || `res`'s `body neutered flag` is set
- The following operations fail when `res.bodyUsed` is set. Otherwise, they acquire the lock of `res.body` and release it when done.
    - `res.arrayBuffer()`
    - `res.blob()`
    - `res.formData()`
    - `res.json()`
    - `res.text()`
- `res.clone()` fail when `res.bodyUsed` is set.
- The following operations fail when `res.bodyUsed` is set. Otherwise, set `res`'s `body neutered flag`, confiscate the underlying source and queue from `res.body` and error it.
    - `e.respondWith(res)`
    - `cache.put(req, res)`

- a `Response`/`Request` with `body neutered flag` is unset but `.body` `"errored"` is considered to be one whose headers were received successfully but body wasn't
- a `Response`/`Request` with `body neutered flag` set is considered to be invalid

## Permanent lock

- `req.bodyUsed` = `req.body` is locked
- The following operations fail when `req.bodyUsed` is set. Otherwise, they acquire the lock of `req.body` and release it when done.
    - `req.arrayBuffer()`
    - `req.blob()`
    - `req.formData()`
    - `req.json()`
    - `req.text()`
- `req.clone()` fail when `req.bodyUsed` is set.
- The following operations fail when `req.bodyUsed` is set. Otherwise, they acquire the lock of `req.body` and never release it.
    - `fetch(req)`
    - `new Request(req)`
    - `cache.put(req, res)`

- `res.bodyUsed` = `res.body` is locked
- The following operations fail when `res.bodyUsed` is set. Otherwise, they acquire the lock of `res.body` and release it when done.
    - `res.arrayBuffer()`
    - `res.blob()`
    - `res.formData()`
    - `res.json()`
    - `res.text()`
- `res.clone()` fail when `res.bodyUsed` is set.
- The following operations fail when `res.bodyUsed` is set. Otherwise, they acquire the lock of `res.body` and never release it.
    - `e.respondWith(res)`
    - `cache.put(req, res)`

- Needs change on the Streams API spec to hold the lock of a stream even after draining it.

**Comment by domenic** [comment]
(A) looks pretty good to me. (B) is not too bad---I could add `getReader({ preventReleaseOnClose: true })`---but it does feel like it is overloading "locked" to mean too many things.

In terms of user-observable consequences, I think (A) allows you to check `req.body.state === "closed"` after `fetch(req)` finishes consuming it, whereas (B) causes it to throw.

**Comment by tyoshino** [comment]
> In terms of user-observable consequences, I think (A) allows you to check req.body.state === "closed" after fetch(req) finishes consuming it, whereas (B) causes it to throw.

In my draft (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71971020), `req.body.state` stays to be `"errored"` even after `fetch(req)` finishes. Do you suggest that we should change it from `"errored"` to `"closed"` once `fetch(req)` finishes?

**Comment by domenic** [comment]
Oh, sorry, I didn't read them carefully enough. Now I am confused though. Why don't those operations acquire a lock and read to the end? They can still set a flag (maybe "body neutered" is not the right name) that impacts bodyUsed.

**Comment by tyoshino** [comment]
For `new Request(req)`'s case, based on Anne's comment (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-70629127), I made it mark the old `req` unavailable permanently. To represent that concept also on `req.body`, I made `req.body` `"errored"` in (A) of https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71971020. I thought it would look more consistent if `req.body` is also made to something indicating unsable-ness (errored or locked (B)) as well as `fetch(req)`, `new Request(req)` and `cache.put(req, res)` are prohibited.

**Comment by domenic** [comment]
I would think in all these cases conceptually what is happening is that the stream is being locked-then-consumed while it is copied over into the new request, or fed to the network, or put into the cache. "waiting" while being consumed, then "closed" after it's done being consumed, makes the most sense to me.

**Comment by tyoshino** [comment]
I'd like to confirm if your opinion is the same as Anne regarding the following question:

`fetch(req)`, `new Request(req)` and `cache.put(req, res)` fails if any of `fetch(req)`, `new Request(req)` and `cache.put(req, res)` has happened even after the operation is finished.

**Comment by tyoshino** [comment]
I got some offline comment about politically incorrectness of naming. I've renamed it to "body passed".

**Comment by tyoshino** [comment]
Copied https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-71971020 to
https://github.com/tyoshino/streams_integration/blob/master/FetchBodyPrecondition.md and added two more ideas (A)' and (C). Domenic, which is depicting your idea?

**Comment by tyoshino** [comment]
Sorry. (A)' was not depicting Domenic's https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-72158403. Added (A)'' which should be depicting it more correctly.

**Comment by tyoshino** [comment]
BTW, Yutaka questioned what we should do with the ReadableStream when e.g. `.text()` fails due to invalid UTF-8 data. We came up with the following options:

1. stays `"readable"` with unconsumed data (or stays `"closed"` if the UTF-8 decoder has already drained all bytes)
1. becomes `"closed"` after draining unconsumed data
1. becomes `"errored"`

**Comment by domenic** [comment]
> I'd like to confirm if your opinion is the same as Anne regarding the following question:
>
> fetch(req), new Request(req) and cache.put(req, res) fails if any of fetch(req), new Request(req) and cache.put(req, res) has happened, even after the operation is finished.

I agree with this.

> Domenic, which is depicting your idea?

(A)'' depicts my idea.

> BTW, Yutaka questioned what we should do with the ReadableStream when e.g. .text() fails due to invalid UTF-8 data. We came up with the following options:

I think 1 or 2 are reasonable and I don't have a strong preference between them. 1 would be text() doing nothing, and 2 would be text() calling `this.body.cancel(new TypeError("couldn't decode chunk!"))`. Maybe since we already think of text() as saying "I want all of the remaining data", 2 is best.

**Comment by annevk** [comment]
utf-8 would not suddenly fail to decode if you hit a trailing byte. You would just get a U+FFFD. I don't think we want to use a fatal decoder.

**Comment by tyoshino** [comment]
OK. So, `text()` uses a replacement mode decoder, so it doesn't fail halfway. In `json()`, we may detect a fatal error in the middle of the data. Is such a case, maybe we could call `cancel()` as Domenic suggested. Similar handling for `formData()`, too?

**Comment by annevk** [comment]
I'm not sure how multipart/form-data is normally processed. I guess we don't parse it ourselves normally so making it a fatal error is fine if we keep doing that consistently.

**Comment by yutakahirano** [comment]
Just a nitpick, but you don't have to have an entry for `fetch(req)` in https://github.com/tyoshino/streams_integration/blob/master/FetchBodyPrecondition.md. It is included in `new Request(req)` case (at least in the current [spec](https://fetch.spec.whatwg.org/#dom-global-fetch), and it seems nobody in this thread wants to change that).

**Comment by tyoshino** [comment]
Yutaka: Thanks. Done https://github.com/tyoshino/streams_integration/commit/723bb07c03fea04a2f1e7072739a753c55ffa8d1

**Comment by tyoshino** [comment]
Thanks, Anne, Domenic. Let's have `json()`, etc. cancel() the stream on any fatal error.

----

Re: Domenic (https://github.com/slightlyoff/ServiceWorker/issues/452#issuecomment-72267929),

OK. Let's proceed with that plan. I've moved (A)'' to the top in https://github.com/tyoshino/streams_integration/blob/master/FetchBodyPrecondition.md.

**Comment by annevk** [comment]
Except `text()`, right? That is, it should not `cancel()` on incorrect bytes.

**Comment by tyoshino** [comment]
Yes. As you said, `text()` just replaces incorrect bytes, so it never fails and therefore is never required to do `cancel()`.

**Comment by tyoshino** [comment]
s/never fails/never fails as long as body doesn't become `"errored"/

**Comment by yutakahirano** [comment]
Thank you @tyoshino  and others, I will update the draft.

**Comment by wanderview** [comment]
What is the type of the "chunk" provided by Fetch body ReadableStream objects?  Is it an ArrayBuffer?  The streams spec says a chunk can be of any type, but we should probably define what Fetch body streams return explicitly.  (Sorry if this is already defined and I missed it.  Its a bit hard to follow the overall effort.)

**Comment by yutakahirano** [comment]
Yes, `read()` returns an ArrayBuffer. I will describe that in the draft.

**Comment by yutakahirano** [comment]
Hi!

Sorry that it took so long, but I updated the [draft] (https://github.com/yutakahirano/fetch-with-streams).
Please let me know if it has any errors regarding bodyUsed, passed flag, locked flag, etc.

**Comment by annevk** [comment]
I suggest any further review is done in that repository until the merge happens (at which point we should switch to https://github.com/whatwg/fetch I think for anything new).

**Comment by xgqfrms-GitHub** [comment]
# TypeError: body stream already read at FetchDatas.fetch.then ???

> what.s wrong with this? Anybody can help?

![promise-bug](https://user-images.githubusercontent.com/18028768/32539857-ca2567e6-c4a5-11e7-8571-c86574a13346.png)
